### PR TITLE
Fixes implicit and explicit generics in procedures

### LINKED ISSUE #17212: Generics issue
Yesterday I tried to save some lines of code in gen.nim file by using one more generic parameter.

Was no problem to join proc glistObjects2seq() and proc gslistObjects2seq() by making parameter l generic, but it does not work for 

https://github.com/StefanSalewski/gintro/blob/master/tests/gen.nim#L2910

Issue is related to this:

```nim
proc t[T](x: T; z: int | float): seq[T] =
  result.add(x & $z)

var y = t[string]("Hallo", 2.0)

proc t2[T](z: int | float): seq[T] =
  result.add($z)

var y2 = t2[string](2.0)
```

/tmp/hhh/t.nim(9, 10) Error: cannot instantiate: 't2[string]'; got 1 typeof(s) but expected 2

/tmp/hhh/t.nim(4, 9) Error: cannot instantiate: 't[string]'; got 1 typeof(s) but expected 2

### PR DESCRIPTION
**Pull request by beef331**
As far as I can tell(searched all issues marked generics) it closes: #17212 #15622 #11152 #4688 #4164

### CODE REVIEW

File: compiler/semexprs.nim [hunk 1]
```diff
@@ -1568,11 +1568,28 @@ proc semSubscript(c: PContext, n: PNode, flags: TExprFlags): PNode =
       of skProc, skFunc, skMethod, skConverter, skIterator:
         # type parameters: partial generic specialization
         n[0] = semSymGenericInstantiation(c, n[0], s)
-        result = explicitGenericInstantiation(c, n, s)
-        if result == n:
-          n[0] = copyTree(result)
-        else:
-          n[0] = result
+        block inferredGenericCheck:
```
Comment (Araq) [inline compiler/semexprs.nim]: Extract it into a helper proc please.

**Comment by beef331** [comment]
Further investigation shows this does enable behaviour but the behaviour is broken. The implicit generics of the same typeclass are always considered the same. Such that the following is valid:
```nim
proc doStuff[Y](a: SomeInteger, b: Y, c: SomeInteger): c.type = discard
assert doStuff[string](100, "hello", 100) is int
assert doStuff[string](100u8, "hello", 255) is uint8 # should be int
assert doStuff[string](100u32, "hello", 355u16) is uint32 # should be uint16
```

File: compiler/semexprs.nim [hunk 2]
```diff
@@ -1493,6 +1493,30 @@ proc semDeref(c: PContext, n: PNode): PNode =
...
+  ## otherwise returns n.
+  let
+    neededGenParams = s.ast[genericParamsPos].len
+    heldGenParams = n.len - 1
+    implicitParams = block:
+      var res = 0
+      for x in s.ast[genericParamsPos]:
+        if tfImplicitTypeParam in x.typ.flags:
+          inc res
+      res
```
Comment (Araq) [inline compiler/semexprs.nim]:
```suggestion
  var implicitParams = 0
  for x in s.ast[genericParamsPos]:
    if tfImplicitTypeParam in x.typ.flags:
      inc implicitParams
```

**Comment by Araq** [comment]
Open/closed to force a re-check against patched testament and important packages.

**Pull request merged**